gdx/src/com/badlogic/gdx/utils/ObjectMap.java

```diff
@@ -1,846 +1,779 @@
+	 * be an expensive operation.
+	 *
+	 * @param identity If true, uses == to compare the specified value with values in the map. If false, uses
+	 *                 {@link #equals(Object)}.
+	 */
+	public boolean containsValue (Object value, boolean identity) {
+		V[] valueTable = this.valueTable;
+		if (value == null) {
+			K[] keyTable = this.keyTable;
+			for (int i = valueTable.length; i-- > 0; )
+				if (keyTable[i] != null && valueTable[i] == null)
+					return true;
+		} else if (identity) {
+			for (int i = valueTable.length; i-- > 0; )
+				if (valueTable[i] == value)
+					return true;
+		} else {
+			for (int i = valueTable.length; i-- > 0; )
+				if (value.equals(valueTable[i]))
+					return true;
+		}
+		return false;
+	}
+
+	public boolean containsKey (K key) {
+		return locateKey(key) != -1;
+	}
+
+	/**
+	 * Returns the key for the specified value, or null if it is not in the map. Note this traverses the entire map and compares
+	 * every value, which may be an expensive operation.
+	 *
+	 * @param identity If true, uses == to compare the specified value with values in the map. If false, uses
+	 *                 {@link #equals(Object)}.
+	 */
+	public K findKey (Object value, boolean identity) {
+		V[] valueTable = this.valueTable;
+		if (value == null) {
+			K[] keyTable = this.keyTable;
+			for (int i = valueTable.length; i-- > 0; )
+				if (keyTable[i] != null && valueTable[i] == null)
+					return keyTable[i];
+		} else if (identity) {
+			for (int i = valueTable.length; i-- > 0; )
+				if (valueTable[i] == value)
+					return keyTable[i];
+		} else {
+			for (int i = valueTable.length; i-- > 0; )
+				if (value.equals(valueTable[i]))
+					return keyTable[i];
+		}
+		return null;
+	}
+
+	/**
+	 * Increases the size of the backing array to accommodate the specified number of additional items. Useful before adding many
+	 * items to avoid multiple backing array resizes.
+	 */
+	public void ensureCapacity (int additionalCapacity) {
+		int sizeNeeded = size + additionalCapacity;
+		if (sizeNeeded >= threshold)
+			resize(MathUtils.nextPowerOfTwo((int)Math.ceil(sizeNeeded / loadFactor)));
+	}
+
+	final void resize (int newSize) {
+		int oldCapacity = keyTable.length;
+		threshold = (int)(newSize * loadFactor);
+		mask = newSize - 1;
+		shift = Long.numberOfLeadingZeros(mask);
+
+		K[] oldKeyTable = keyTable;
+		V[] oldValueTable = valueTable;
+
+		keyTable = (K[])new Object[newSize];
+		valueTable = (V[])new Object[newSize];
+
+		int oldSize = size;
+		size = 0;
+		if (oldSize > 0) {
+			for (int i = 0; i < oldCapacity; i++) {
+				K key = oldKeyTable[i];
+				if (key != null)
+					putResize(key, oldValueTable[i]);
+			}
+		}
+	}
+
+	public int hashCode () {
+		int h = size;
+		K[] keyTable = this.keyTable;
+		V[] valueTable = this.valueTable;
+		for (int i = 0, n = keyTable.length; i < n; i++) {
+			K key = keyTable[i];
+			if (key != null) {
+				h ^= key.hashCode();
+
+				V value = valueTable[i];
```

Comment:
use reference instead of variable creation

Reply:
Using the local variable to do a null check is safer and better practice.

Reply:
So here you're saying to use this?
```java
if(valueTable[i] != null) h+= valueTable[i].hashCode();
```

Reply:
Correct, we'd benchmark it,
I'm not very sure if this is faster, but with only few accesses it should be faster

Reply:
> Using the local variable to do a null check is safer and better practice.

Why it's saver?